### PR TITLE
[SPARK-47577][CORE][PART2] Migrate logError with variables to structured logging framework 

### PR DESCRIPTION
**Pull request by gengliangwang**
### What changes were proposed in this pull request?

Migrate logError with variables of core module to structured logging framework. This is part2 which transforms the logError entries of the following API
```
def logError(msg: => String, throwable: Throwable): Unit
```
to
```
def logError(entry: LogEntry, throwable: Throwable): Unit
```

migration Part1 was in https://github.com/apache/spark/pull/45834
### Why are the changes needed?

To enhance Apache Spark's logging system by implementing structured logging. 
### Does this PR introduce _any_ user-facing change?

Yes, Spark core logs will contain additional MDC

### How was this patch tested?

Compiler and scala style checks, as well as code review.

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Comment by gengliangwang** [comment]
cc @dtenedor @panbingkun @pan3793 @itholic as well

File: core/src/main/scala/org/apache/spark/util/Utils.scala [hunk 1]
```diff
@@ -1276,11 +1276,13 @@ private[spark] object Utils
       case t: Throwable =>
         val currentThreadName = Thread.currentThread().getName
         if (sc != null) {
-          logError(s"uncaught error in thread $currentThreadName, stopping SparkContext", t)
+          logError(
+            log"uncaught error in thread ${MDC(THREAD, currentThreadName)}, stopping SparkContext",
```
Comment (panbingkun) [inline core/src/main/scala/org/apache/spark/util/Utils.scala:1280]: Will we have `THREAD_ID` in the future? `THREAD_NAME`?

File: core/src/main/scala/org/apache/spark/util/Utils.scala [hunk 2]
```diff
@@ -1276,11 +1276,13 @@ private[spark] object Utils
...
+            t)
           sc.stopInNewThread()
         }
         if (!NonFatal(t)) {
-          logError(s"throw uncaught fatal error in thread $currentThreadName", t)
+          logError(log"throw uncaught fatal error in thread ${MDC(THREAD, currentThreadName)}", t)
```
Comment (panbingkun) [inline core/src/main/scala/org/apache/spark/util/Utils.scala:1285]: dito

File: core/src/main/scala/org/apache/spark/storage/PushBasedFetchHelper.scala [hunk 3]
```diff
@@ -181,8 +183,9 @@ private class PushBasedFetchHelper(
 
       override def onFailure(shuffleId: Int, shuffleMergeId: Int, reduceId: Int,
           exception: Throwable): Unit = {
-        logError(s"Failed to get the meta of push-merged block for ($shuffleId, $reduceId) " +
-          s"from ${req.address.host}:${req.address.port}", exception)
+        logError(log"Failed to get the meta of push-merged block for " +
```
Comment (panbingkun) [inline core/src/main/scala/org/apache/spark/storage/PushBasedFetchHelper.scala:186]: It's not related to this PR, do we need `shuffleMergeId` in the log?

File: core/src/main/scala/org/apache/spark/executor/Executor.scala [hunk 4]
```diff
@@ -661,9 +661,10 @@ private[spark] class Executor(
           // uh-oh.  it appears the user code has caught the fetch-failure without throwing any
           // other exceptions.  Its *possible* this is what the user meant to do (though highly
           // unlikely).  So we will log an error and keep going.
-          logError(s"$taskName completed successfully though internally it encountered " +
-            s"unrecoverable fetch failures!  Most likely this means user code is incorrectly " +
-            s"swallowing Spark's internal ${classOf[FetchFailedException]}", fetchFailure)
+          logError(log"${LogMDC(TASK_NAME, taskName)} completed successfully though internally " +
+            log"it encountered unrecoverable fetch failures!  Most likely this means user code " +
```
Comment (panbingkun) [inline core/src/main/scala/org/apache/spark/executor/Executor.scala:665]: Can we correct the original appearance of the two whitespaces before `Most`?

File: core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala [hunk 5]
```diff
@@ -794,8 +796,10 @@ private[deploy] class Worker(
           case Failure(t) =>
             val failures = executorStateSyncFailureAttempts.getOrElse(fullId, 0) + 1
             if (failures < executorStateSyncMaxAttempts) {
-              logError(s"Failed to send $newState to Master $masterRef, " +
-                s"will retry ($failures/$executorStateSyncMaxAttempts).", t)
+              logError(log"Failed to send ${MDC(EXECUTOR_STATE_CHANGED, newState)}" +
```
Comment (panbingkun) [inline core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala:799]: I'm not sure if `EXECUTOR_STATE_CHANGE` or `EXECUTOR_STATE` is better

**Review by dongjoon-hyun — COMMENTED** [review]
nit, the PR description seems to be wrong to me, @gengliangwang . `logInfo` is a typo, right? :) 

Screenshot 2024-04-05 at 07 57 29

File: common/utils/src/main/scala/org/apache/spark/internal/LogKey.scala [hunk 6]
```diff
@@ -33,40 +34,56 @@ object LogKey extends Enumeration {
...
   val REASON = Value
+  val REDUCE_ID = Value
   val REMOTE_ADDRESS = Value
   val RETRY_COUNT = Value
   val RPC_ADDRESS = Value
+  val SHUFFLE_BLOCK_INFO = Value
   val SHUFFLE_ID = Value
+  val SHUFFLE_MERGE_ID = Value
   val SIZE = Value
+  val SLEEP_TIME_SECONDS = Value
```
[the remark this answers is not in the document]
Comment (gengliangwang) [reply common/utils/src/main/scala/org/apache/spark/internal/LogKey.scala:86]: We can handle this when there is SLEEP_TIME_MILLISECONDS

**Comment by gengliangwang** [comment]
@dongjoon-hyun Yes, updated :)

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM from my side.

**Comment by dongjoon-hyun** [comment]
Thank you! Please add the existing comments first, @gengliangwang .

File: core/src/main/scala/org/apache/spark/util/Utils.scala (hunk 1 above)
Comment (gengliangwang) [reply core/src/main/scala/org/apache/spark/util/Utils.scala:1280]: Good point, I will update this.

File: core/src/main/scala/org/apache/spark/executor/Executor.scala (hunk 4 above)
Comment (gengliangwang) [reply core/src/main/scala/org/apache/spark/executor/Executor.scala:665]: Sure.

File: core/src/main/scala/org/apache/spark/storage/PushBasedFetchHelper.scala (hunk 3 above)
Comment (gengliangwang) [reply core/src/main/scala/org/apache/spark/storage/PushBasedFetchHelper.scala:186]: Well let's just keep it.

**Review by xinrong-meng — APPROVED** [review]

**Comment by xinrong-meng** [comment]
nit: pr description says "part1" which is inconsistent with the pr title :)

**Comment by gengliangwang** [comment]
@panbingkun @dongjoon-hyun @xinrong-meng Thanks for the review, merging to master

**Pull request closed without merging**